File: 13-roman-to-integer/roman-to-integer.ts
function romanToInt(s: string): number {
    let mp = new Map<string, number>([
        ['I', 1],
        ['V', 5],
        ['X', 10],
        ['L', 50],
        ['C', 100],
        ['D', 500],
        ['M', 1000],
    ]);
  
    let ans = mp.get(s.charAt(s.length - 1));

    for(let i = s.length - 2; i >= 0; i--) {
        if(mp.get(s.charAt(i)) < mp.get(s.charAt(i + 1))) {
            ans -= mp.get(s.charAt(i));
        }
        else ans += mp.get(s.charAt(i));
    }
    return ans;
};

File: 13-roman-to-integer/roman-to-integer.java
class Solution {
    public int romanToInt(String s) {
        Map<Character , Integer> mp = new HashMap<>();
        int ans = 0;
        mp.put('I',1);
        mp.put('V',5);
        mp.put('X',10);
        mp.put('L',50);
        mp.put('C',100);
        mp.put('D',500);
        mp.put('M',1000);
        int n = s.length();

        for(int i = n - 1; i >= 0; i --) {
            int curCharValue = mp.get(s.charAt(i));
            if(i == n - 1) ans += curCharValue;
            else {
                int nextCharValue = mp.get(s.charAt(i + 1));
                if(curCharValue < nextCharValue) ans -= curCharValue;
                else ans += curCharValue;
            }
        }
        return ans;
    }
}